lib/pbench/agent/tool_meister.py

```diff
@@ -844,32 +881,37 @@ def __enter__(self):
         version, seqno, sha1, hostdata = collect_local_info(self.pbench_install_dir)
 
         tool_installs = {}
-        for name, tool_opts in sorted(self._tools.items()):
+        for name, tool_opts in sorted(self._params.tools.items()):
             tklass = self._tool_name_class_mappings.get(name, TransientTool)
             try:
                 tool = tklass(
                     name,
                     tool_opts,
                     pbench_install_dir=self.pbench_install_dir,
-                    tool_dir=self._tool_dir,
                     logger=self.logger,
                 )
-                # FIXME - consider running these in parallel.
-                tool_installs[name] = tool.install()
-            except Exception:
+                # Note - each ".install()" method runs sequentially.
+                res = tool.install()
+            except Exception as exc:
                 self.logger.exception("Failed to run tool %s install check", name)
-                tool_installs[name] = (-42, "internal-error")
-        self._failed_tools = {}
-        for name, res in tool_installs.items():
-            if res[0] != 0:
-                self.logger.debug("Recording failed tool, %s", name)
-                self._failed_tools[name] = self._tools[name]
-                del self._tools[name]
+                res = InstallationResult(
+                    returncode=-42, output=f"internal-error: {exc}"
+                )
```

Comment:
It's a bit redundant to include the `Exception` text here since you're implicitly chaining the original `cause` on your `InstallationResult`; the stack trace knows all, tells all. (Plus, you're logging `.exception` which will also have the full stack trace.)

Reply:
I was asked to make this change and capitulated.  I feel like a ping-pong ball.  =)

Reply:
You mean adding the `{exc}` in the `InstallationResult` message string? That's all I was talking about. Nevermind.

Reply:
I am not in favor of sending the exception string representation back in the list of results of tool installs.

First, it is "internal" to the operation of the Tool Meister, and the most of the errors that will be recorded for failed tool installs are consumed by the users who setup the systems to run the pbench-agent.  Those people are not going to have a good idea about what to do with an internal error in the Tool Meister sub-system, nor would we want to require them to know what to do.

Second, the exception log entry will have already been sent back to the Tool Data Sink and stored in the `tm.logs` file, so a pbench engineer debugging a problem would still have it.

So I have convinced myself to remove it.

Reply:
> the original cause should be chained to the new exception, and including the exception text seems redundant.

Yeah, the `logging.exception()` output would contain the whole traceback, including any previous cause; however, the stringized exception object that I'd like to see attached to the result message would only contain the current exception's text -- the traceback and any chained exceptions wouldn't be included.

> We're doing a `raise` inside an `except` without `from None` 

I'm not sure where that's happening; I think this `except` clause would be catching `subprocess` invocation failures.